app/components/graph-editor/drawable-interfaces.ts

```diff
@@ -0,0 +1,398 @@
+}
+
+/**
+ * isDrawableNode  
+ *   Typeguard for drawable nodes.
+ */
+export function isDrawableNode(obj: any): obj is DrawableNode {
+    return isItThat(obj, NODE_DEFAULTS);
+}
+
+/**
+ * isItThat  
+ *   Checks if `it` has the same members as `that`.
+ */
+function isItThat(it: any, that: any): boolean {
+    if (!it)
+        return false;
+    let result = true;
+    for (let p in that) {
+        result =
+            result &&
+            (p in it) &&
+            (typeof that[p] === typeof it[p]);
+        if (!result)
+            return false;
+    }
+    return result;
+}
+
+/**
+ * isEdgeOverlapped  
+ *   Checks if two edges are overlapping.  
+ *   It is assumed that the edge being checked does not have the same source
+ *   and destination nodes.
+ */
+export function isEdgeOverlapped(e: DrawableEdge, nodeEdges: Map<DrawableNode, Set<DrawableEdge>>): boolean {
+    for (const edge of (nodeEdges.get(e.source as DrawableNode) as Set<DrawableEdge>)) {
+        if (
+            e !== edge &&
+            e.source === edge.destination &&
+            e.destination === edge.source
+        )
+            return true;
+    }
+    for (const edge of (nodeEdges.get(e.destination as DrawableNode) as Set<DrawableEdge>)) {
+        if (
+            e !== edge &&
+            e.source === edge.destination &&
+            e.destination === edge.source
+        )
+            return true;
+    }
+    return false;
+}
+
+/**
+ * getOverlappedEdges  
+ *   Gets the set of edges that overlap with a given edge.  
+ *   It is assumed that the edge being checked does not have the same source
+ *   and destination nodes.
+ */
+export function getOverlappedEdges(e: DrawableEdge, nodeEdges: Map<DrawableNode, Set<DrawableEdge>>): Set<DrawableEdge> {
+    let overlapped = new Set<DrawableEdge>();
+    for (const edge of (nodeEdges.get(e.source as DrawableNode) as Set<DrawableEdge>)) {
+        if (
+            e !== edge &&
+            e.source === edge.destination &&
+            e.destination === edge.source
+        )
+            overlapped.add(edge);
+    }
+    for (const edge of (nodeEdges.get(e.destination as DrawableNode) as Set<DrawableEdge>)) {
+        if (
+            e !== edge &&
+            e.source === edge.destination &&
+            e.destination === edge.source
+        )
+            overlapped.add(edge);
+    }
+    return overlapped;
+}
+
+/**
+ * cloneEdge  
+ *   Creates a cloned edge.
+ */
+export function cloneEdge(e: DrawableEdge): DrawableEdge {
+    let clone = new DefaultEdge();
+    clone.source = e.source;
+    clone.destination = e.destination;
+    clone.showSourceArrow = e.showSourceArrow;
+    clone.showDestinationArrow = e.showDestinationArrow;
+    clone.color = e.color;
+    clone.lineStyle = e.lineStyle;
+    clone.lineWidth = e.lineWidth;
+    clone.label = e.label;
+    return clone;
+}
+
+/**
+ * cloneNode  
+ *   Creates a cloned node.
+ */
+export function cloneNode(n: DrawableNode): DrawableNode {
```

Comment:
In my opinion, this should be replaced with a generic cloning function if possible or be a method of the DrawableNode class. i.e., it should only depend on the contents of DrawableNode if it is a method else it should generically do a deep copy.